### PR TITLE
Add CLN

### PR DESCRIPTION
**Pull request by m3dwards**
Adds CLN as a lightning node option

### CODE REVIEW

**Review by willcl-ark — COMMENTED** [review]
Thanks for this PR, just taken a brief surface-level skim.

It seems to me that we have ended up with `lnnode.py` just including a tonne of `if...else` statements to support the two implementations. If we ever decide to add LDK this will only get worse. Perhaps we should, before merging this, create an `LNNode` base class, and subclass `LndNode` and `CLNNode` as we did previously for the two backends?

That way we keep them clean, and make adding new nodes much easier in the future?

It would also let us abstract away things like: https://github.com/bitcoin-dev-project/warnet/pull/355/files#diff-0bfb3fbf6aa1bb068dffe115137cfec50240a3e8713463149f47d8e1354ceef1R100-R107 too...?

File: test/graph_test.py [hunk 1]
```diff
@@ -68,20 +68,42 @@
...
+    )
+
+
+def policy_match(pol1, pol2):
+    return (
+        max(int(pol1["time_lock_delta"]), 18) == max(int(pol2["time_lock_delta"]), 18)
+        and max(int(pol1["min_htlc"]), 1) == max(int(pol2["min_htlc"]), 1)
+        and pol1["fee_base_msat"] == pol2["fee_base_msat"]
+        and pol1["fee_rate_milli_msat"] == pol2["fee_rate_milli_msat"]
+    )
```
Comment (pinheadmz) [inline test/graph_test.py:90]: Aren't these defined in utils.py? Do they need to be stubbed here?

File: test/ln_test.py [hunk 2]
```diff
@@ -97,7 +100,7 @@ def check_invoices(index):
 
 print("\nEngaging simln")
 activity = [{"source": "ln-0", "destination": node2pub, "interval_secs": 1, "amount_msat": 2000}]
-base.warcli(f"network export --exclude=[1] --activity={json.dumps(activity).replace(' ', '')}")
+base.warcli(f"network export --exclude=[1,3] --activity={json.dumps(activity).replace(' ', '')}")
```
Comment (pinheadmz) [inline test/ln_test.py]: The `exclude` option is intended for the attackathon to make the target node forwarding-only: https://github.com/carlaKC/attackathon/issues/19

simln should support cln ok, is there something else Im missing?

File: test/data/ln.graphml [hunk 3]
```diff
@@ -38,22 +38,33 @@
         <data key="ln_config">--bitcoin.timelockdelta=33</data>
     </node>
     <node id="3">
+        <data key="version">27.0</data>
+        <data key="bitcoin_config">-uacomment=w2</data>
+        <data key="ln">cln</data>
+        <data key="ln_config">--bitcoin.timelockdelta=33</data>
```
Comment (pinheadmz) [inline test/data/ln.graphml]: this is an arg passed to the ln container -- does CLN use the same syntax as lnd for this?

**Review by pinheadmz — COMMENTED** [review]
Reviewed and tested, really cool to see cln in there. I was able to get invoices and make payments in the 3 node graph. I agree with Will though, I think the implementations should be subclasses of `LNNode`. If that sounds like too much work for now we can address it in a follow-up. Functionally this PR works great for now.

I also wonder how we should handle `warcli lncli... ` commands. Like, we should we abstract the LN commands such that all lnd commands work for cln and ldk as well?

File: test/graph_test.py (hunk 1 above)
Comment (m3dwards) [reply test/graph_test.py:90]: I really hoped you wouldn't notice this but you are too smart for me.

The difference is that in in utils.py it now works with LNChannel objects which are returned from helper functions from LNNode. LNChannel is an abstraction over channels objects as returned by LND or CLN.

Why I couldn't use that in the test is that the test framework only uses the RPC which in this test is returning specific JSON from LND. I couldn't think of a quick solution and so added these helper functions here to keep the test passing. Ideally I think more and more things should move to use the abstractions and the CLI should provide functions for common lightning functions.

File: test/ln_test.py (hunk 2 above)
Comment (m3dwards) [reply test/ln_test.py]: Could you perhaps help me get the test passing without `,3`?

File: test/data/ln.graphml (hunk 3 above)
Comment (m3dwards) [reply test/data/ln.graphml]: Good spot, no it is not.

**Comment by m3dwards** [comment]
> Perhaps we should, before merging this, create an `LNNode` base class, and subclass `LndNode` and `CLNNode` as we did previously for the two backends

Agreed. Much nicer.

**Comment by m3dwards** [comment]
@willcl-ark @pinheadmz switched to lightning subclasses.

File: test/graph_test.py [hunk 4]
```diff
@@ -68,20 +68,42 @@
...
+
+
 print("Ensuring warnet LN channel policies match imported JSON description")
 with open(json_file_path) as file:
     actual = json.loads(base.warcli("lncli 0 describegraph"))["edges"]
+    # json.dump(actual, file.writable(), indent=2)
```
Comment (willcl-ark) [inline test/graph_test.py]: Accidental inclusion?

**Review by willcl-ark — APPROVED** [review]
Looks and seems to work great! nice work.

I left a single comment here, and also prepared 3 additional commits which you may be interested in taking here, or in a followup?

https://github.com/bitcoin-dev-project/warnet/compare/main...willcl-ark:warnet:add-cln-review

If you'd prefer followup, we can merge when @pinheadmz question about the `--exclude` option is resolved IMO

**Comment by m3dwards** [comment]
> 3 additional commits which you may be interested in taking here

Taken

File: test/ln_test.py (hunk 2 above)
Comment (m3dwards) [reply test/ln_test.py]: Should be sorted now

**Comment by pinheadmz** [comment]
#408

**Pull request closed without merging**